py/client/pydeephaven/session.py

```diff
@@ -226,6 +235,28 @@ def __exit__(self, exc_type, exc_val, exc_tb):
     def __del__(self):
         self.close()
 
+    def update_metadata(self, metadata: Iterable[Tuple[bytes, bytes]]):
+        for header_tuple in metadata:
+            if header_tuple[0] == "authorization":
+                self._auth_header_value = bytes(header_tuple[1], 'ascii')
+                break
+
+    def wrap_rpc(self, stub_call, *args, **kwargs):
+        if 'metadata' in kwargs:
+            raise DHError('Internal error: "metadata" in kwargs not supported in wrap_rpc.')
+        kwargs["metadata"] = self.grpc_metadata
+        response, call = stub_call.with_call(*args, **kwargs)
+        self.update_metadata(call.initial_metadata())
+        return response
+
+    def wrap_bidi_rpc(self, stub_call, *args, **kwargs):
+        if 'metadata' in kwargs:
+            raise DHError('Internal error: "metadata" in kwargs not supported in wrap_bidi_rpc.')
+        kwargs["metadata"] = self.grpc_metadata
+        response = stub_call(*args, **kwargs)
+        self.update_metadata(response.initial_metadata())
```

Comment:
One irritation about expressing it this way is that the server could send back headers right away (with an updated auth header), then block for a bit to get work done before sending the payload. As written, I believe the following could happen:
1. Make gRPC call A, with token 1
2. gRPC call A sends headers, with token 1 still
3. Make gRPC call B, with token 1
4. gRPC call B sends headers with refreshed token 2
5. gRPC call B sends response
6. update_metadata is called with token 2, replaces our existing token 1
7. gRPC call A sends response
8. update_metadata is called with token 1, replaces our existing token 2
9. next gRPC call to go out uses token 1, which might have now expired

We would need the stars to align against us badly to fail twice and lose the session (where call B was the keepalive, etc), but it does appear possible. Using a gRPC interceptor guards against this, by passing headers before any payload.

Reply:
Did not think of that.   At first read this sounds like a very valid concern (damn).
You and I discussed over slack, it is possible we can avoid this by implementing `wrap_rpc` to use the `future` API for non-bidi stubs (we discussed that the wrap_bidi_rpc is already doing the right thing, meaning this concern does not aply there).  We agreed I will experiment with the code to try to move `wrap_rpc` to use the gRPC `future` API to stubs and come back.

Thanks for noticing this, this was an oversight.

Reply:
Done, the future API was not hard to use, and the result is much like what we were already doing with `wrap_bidi_rpc`, PTAL.